File: src/structure/java/LinkedList.java
package structure.java;

import java.util.List;

public class LinkedList<T> {
    private class Node<T> {
        public T element;
        public Node next;

        public Node(T e) {
            element = e;
        }

        public Node(T e, Node n) {
            element = e;
            next = n;
        }

        public Node() {
        }

        @Override
        public String toString() {
            return element.toString();
        }
    }

    private Node dummyHead = new Node();
    int size;

    public LinkedList() {
        size = 0;
    }

    public LinkedList(T e) {
        size = 1;
        dummyHead.next = new Node<>(e);
    }

    public LinkedList(List<T> list) {
        size = list.size();
        if (size == 0) {
            return;
        }
        dummyHead.next = new Node<>(list.get(0));
        Node now = dummyHead.next;
        for (int i = 1; i < list.size(); i++) {
            now.next = new Node<>(list.get(i));
            now = now.next;
        }
    }

    public int getSize() {
        return size;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public void addFist(T e) {
        add(0, e);
    }

    public void add(int index, T e) {
        if (index < 0 || index > size)
            throw new IndexOutOfBoundsException();
        Node now = dummyHead;
        for (int i = 0; i < index; i++)
            now = now.next;
        now = new Node(e, now.next);
        size++;
    }

    public void addLast(T e) {
        add(size, e);
    }

    public T get(int index) {
        return (T) getNode(index).element;
    }

    public void set(int index) {
        Node now = getNode(index);
    }

    private Node getNode(int index) {
        if (index < 0 || index >= size)
            throw new IndexOutOfBoundsException();
        Node now = dummyHead.next;
        for (int i = 0; i < index; i++) {
            now = now.next;
        }
        return now;
    }

    private boolean contains(T e) {
        Node now = dummyHead;
        while (now != null) {
            if (now.element == e) {
                return true;
            }
            now = now.next;
        }
        return false;
    }

    private T remove(int index) {
        if (index < 0 || index >= size)
            throw new IndexOutOfBoundsException();
        Node pre = dummyHead;
        for (int i = 0; i < index; i++) {
            pre = pre.next;
        }
        Node delNode = pre.next;
        pre.next = pre.next.next;
        delNode.next = null;
        size--;
        return (T) delNode.element;
    }

    public T removeFirst() {
        return remove(0);
    }

    public T removeLst() {
        return remove(size - 1);
    }
}

File: src/structure/kotlin/LinkedList.kt
package structure.kotlin

import kotlin.IndexOutOfBoundsException

class LinkedList<T>(vararg data: T) {
    private var dummyHead: Node<T> = Node(null, null)
    private var head: Node<T>?
        get() = dummyHead.next
        set(value) {
            dummyHead.next = value
        }
    var size: Int = data.size
        private set
    val empty
        get() = size == 0

    init {
        if (size != 0) {
            dummyHead.next = Node(data[0])
            var now = head
            for (i in 1 until data.size) {
                now!!.next = Node(data[i])
                now = now.next
            }
        }
    }

    private class Node<T>(var element: T?, var next: Node<T>? = null) {
        override fun toString(): String {
            return element.toString()
        }
    }

    fun addFirst(e: T) {
        head = Node(e, head)
        size++
    }

    fun addLast(e: T) {
        add(e)
    }

    fun add(element: T, index: Int = size) {
        if (index < 0 || index > size)
            throw IndexOutOfBoundsException()
        var now: Node<T> = dummyHead
        for (i in 0 until index)
            now = now.next!!
        now.next = Node(element, now.next)
        size++
    }

    operator fun get(index: Int): T = getNode(index).element!!

    private fun getNode(index: Int): Node<T> {
        return getPre(index).next!!
    }

    fun getFirst() = get(0)
    fun getLast() = get(size - 1)

    private fun getPre(index: Int): Node<T> {
        if (index < 0 || index >= size)
            throw IndexOutOfBoundsException()
        var now: Node<T> = dummyHead
        for (i in 0 until index)
            now = now.next!!
        return now
    }

    operator fun set(index: Int, element: T) {
        val now = getNode(index)
        now.element = element
    }

    fun contains(element: T): Boolean {
        if (empty) return false
        var now = head
        while (now != null) {
            if (now.element!! == element) {
                return true
            }
            now = now.next
        }
        return false
    }

    fun remove(index: Int): T {
        val pre = getPre(index)
        val delNode = pre.next!!
        pre.next = pre.next!!.next
        delNode.next = null
        size--
        return delNode.element!!
    }

    fun removeLast() = remove(size - 1)
    fun removeFirst() = remove(0)

    override fun toString(): String {
        val result = arrayListOf<T>()
        if (empty)
            return "LinkedNode[]"
        var now = head
        while (now != null) {
            result.add(now.element!!)
            now = now.next
        }
        return result.joinToString(" -> ", "LinkedNode[ ", " ]")
    }

}

fun main(args: kotlin.Array<String>) {
    val linkedListKt = LinkedList<Int>(1, 2, 3, 4, 5, 6)
    linkedListKt.add(0,3)
    linkedListKt[3] = 5
    println(linkedListKt.remove(4))
    println(linkedListKt.contains(21))
    println(linkedListKt[3])
    println(linkedListKt)
    linkedListKt.removeFirst()
    println(linkedListKt)
}